### ISSUE TITLE
Do I need to pad?

### ISSUE THREAD

**Issue by timsoraro** [lucidrains/sinkhorn-transformer#8]
Excuse me for the noob question, I have sequences of different lengths and I use:
```python
model = AutoregressiveWrapper(model, ignore_index=PAD_ID, pad_value=PAD_ID)
```

Do I need to pad as in this example or not?
```python
def __getitem__(self, index):
        seq_tokens = self.examples[index]
        input_ids = torch.tensor(seq_tokens, dtype=torch.long)
        input_ids = F.pad(input_ids, (seq_len - len(input_ids), 0), value=self.pad_value)
        return input_ids
```

**Comment by lucidrains** [comment]
@timsoraro oh man, I wished I had finished this before you came! I am enroute to having this be abstracted away just like reformer-pytorch (bringing in the AutoPadder class). I'll try to get it done and tested by tonight

basically, Sinkhorn comes with the drawback that it is fixed to a certain number of buckets, and after toying around with it, I realized it is unable to generalize to different sequence lengths and window sizes. After a lot of experimentation, I think I have prepared a new sorting network that should be able to generalize to multiples of the bucket size, but it requires reframing the code differently, since everything is coded to number of buckets rather than bucket size. So, TLDR, for now you will have to actually pad to the maximum sequence length and set the appropriate input mask :|

I should be able to get this done by tomorrow though! Want me to ping you when it is finished? I've been hoping you would come lol

**Comment by timsoraro** [comment]
Haha, that would be awesome, thanks a lot!

EDIT: R.

**Comment by timsoraro** [comment]
Btw, I sent you an email to your Gmail account (:

**Comment by lucidrains** [comment]
@timsoraro got your email :) I'll try to have this done by end of the week!
<img width="782" alt="Screen Shot 2020-04-15 at 7 34 53 PM" src="https://user-images.githubusercontent.com/108653/79482839-3b197e00-7fc6-11ea-8ac0-de88dcac4251.png">

**Comment by lucidrains** [comment]
@timsoraro I think it is ready in `0.5.1`, ahead of schedule!

**Comment by timsoraro** [comment]
@lucidrains Awesome, going to test (:

Do I still need to pad using this line?
```python
input_ids = F.pad(input_ids, (seq_len - len(input_ids), 0), value=self.pad_value)
```

**Comment by lucidrains** [comment]
@timsoraro no, you shouldn't have to! the AutoRegressive class wraps it all in the Autopadder, which takes care of all padding issues for you

Edit - including masks as well!

**Comment by timsoraro** [comment]
Great (:
